File: mcp_server_erlang/src/mcp_protocol.erl
%%%-------------------------------------------------------------------
%%% @doc MCP Protocol - JSON-RPC 2.0 handling for MCP server
%%%
%%% Handles reading/writing JSON-RPC messages over stdio with
%%% Content-Length framing as required by MCP.
%%%
%%% Copyright (C) 2024 ALGT Project
%%%-------------------------------------------------------------------
-module(mcp_protocol).

-export([
    read_message/0,
    write_response/1,
    write_error/3,
    write_notification/2,
    error_response/3
]).

%% JSON-RPC 2.0 Error Codes
-define(PARSE_ERROR, -32700).
-define(INVALID_REQUEST, -32600).
-define(METHOD_NOT_FOUND, -32601).
-define(INVALID_PARAMS, -32602).
-define(INTERNAL_ERROR, -32603).

-export([
    parse_error/0,
    invalid_request/0,
    method_not_found/0,
    invalid_params/0,
    internal_error/0
]).

parse_error() -> ?PARSE_ERROR.
invalid_request() -> ?INVALID_REQUEST.
method_not_found() -> ?METHOD_NOT_FOUND.
invalid_params() -> ?INVALID_PARAMS.
internal_error() -> ?INTERNAL_ERROR.

%%%-------------------------------------------------------------------
%%% @doc Read a JSON-RPC message from stdin
%%% Messages are framed with Content-Length headers
%%%-------------------------------------------------------------------
-spec read_message() -> {ok, map()} | {error, term()} | eof.
read_message() ->
    case read_headers() of
        eof ->
            eof;
        {error, Reason} ->
            {error, Reason};
        Headers ->
            case proplists:get_value(content_length, Headers) of
                undefined ->
                    {error, missing_content_length};
                Length ->
                    case io:get_chars(standard_io, "", Length) of
                        eof ->
                            eof;
                        {error, Reason} ->
                            {error, Reason};
                        Data ->
                            try
                                Json = jsx:decode(iolist_to_binary(Data), [return_maps]),
                                {ok, Json}
                            catch
                                _:Error ->
                                    {error, {json_parse_error, Error}}
                            end
                    end
            end
    end.

%%%-------------------------------------------------------------------
%%% @doc Read HTTP-style headers until empty line
%%%-------------------------------------------------------------------
-spec read_headers() -> [{atom(), term()}] | eof | {error, term()}.
read_headers() ->
    read_headers([]).

read_headers(Acc) ->
    case io:get_line(standard_io, "") of
        eof ->
            eof;
        {error, Reason} ->
            {error, Reason};
        Line ->
            Trimmed = string:trim(Line, both),
            case Trimmed of
                "" ->
                    Acc;
                _ ->
                    case parse_header(Trimmed) of
                        {ok, Header} ->
                            read_headers([Header | Acc]);
                        ignore ->
                            read_headers(Acc)
                    end
            end
    end.

%%%-------------------------------------------------------------------
%%% @doc Parse a single header line
%%%-------------------------------------------------------------------
-spec parse_header(string()) -> {ok, {atom(), term()}} | ignore.
parse_header(Line) ->
    case string:split(string:lowercase(Line), ":") of
        ["content-length", Value] ->
            try
                Length = list_to_integer(string:trim(Value)),
                {ok, {content_length, Length}}
            catch
                _:_ -> ignore
            end;
        _ ->
            ignore
    end.

%%%-------------------------------------------------------------------
%%% @doc Write a JSON-RPC response to stdout
%%%-------------------------------------------------------------------
-spec write_response(map()) -> ok.
write_response(Response) ->
    JsonBin = jsx:encode(Response),
    Length = byte_size(JsonBin),
    io:format(standard_io, "Content-Length: ~B\r\n\r\n~s", [Length, JsonBin]),
    ok.

%%%-------------------------------------------------------------------
%%% @doc Write a JSON-RPC error response
%%%-------------------------------------------------------------------
-spec write_error(term(), integer(), binary()) -> ok.
write_error(Id, Code, Message) ->
    Response = #{
        <<"jsonrpc">> => <<"2.0">>,
        <<"id">> => Id,
        <<"error">> => #{
            <<"code">> => Code,
            <<"message">> => Message
        }
    },
    write_response(Response).

%%%-------------------------------------------------------------------
%%% @doc Write a JSON-RPC notification
%%%-------------------------------------------------------------------
-spec write_notification(binary(), map()) -> ok.
write_notification(Method, Params) ->
    Notification = #{
        <<"jsonrpc">> => <<"2.0">>,
        <<"method">> => Method,
        <<"params">> => Params
    },
    write_response(Notification).

%%%-------------------------------------------------------------------
%%% @doc Create an error response map
%%%-------------------------------------------------------------------
-spec error_response(term(), integer(), binary()) -> map().
error_response(Id, Code, Message) ->
    #{
        <<"jsonrpc">> => <<"2.0">>,
        <<"id">> => Id,
        <<"error">> => #{
            <<"code">> => Code,
            <<"message">> => Message
        }
    }.

File: mcp_servers/prolog/mcp_protocol.pl
%% mcp_protocol.pl
%%
%% JSON-RPC 2.0 protocol handling for MCP server
%%
%% Copyright (C) 2024 ALGT Project

:- module(mcp_protocol, [
    read_jsonrpc_message/2,
    write_jsonrpc_response/2,
    write_jsonrpc_error/4,
    write_jsonrpc_notification/3,
    jsonrpc_error_response/4,
    parse_error/1,
    invalid_request/1,
    method_not_found/1,
    invalid_params/1,
    internal_error/1
]).

:- use_module(library(http/json)).
:- use_module(library(readutil)).

%% JSON-RPC 2.0 Error Codes
parse_error(-32700).
invalid_request(-32600).
method_not_found(-32601).
invalid_params(-32602).
internal_error(-32603).

%% read_jsonrpc_message(+Stream, -Message)
%%
%% Reads a JSON-RPC message from the stream.
%% MCP stdio transport uses newline-delimited JSON (one JSON object per line).

read_jsonrpc_message(Stream, Message) :-
    read_line_to_string(Stream, Line),
    (   Line == end_of_file
    ->  Message = end_of_file
    ;   string_trim(Line, Trimmed),
        Trimmed == ""
    ->  read_jsonrpc_message(Stream, Message)  % skip blank lines
    ;   string_trim(Line, Trimmed),
        catch(
            atom_json_dict(Trimmed, Message, []),
            Error,
            (   log_error("JSON parse error: ~w", [Error]),
                Message = json_error(Error)
            )
        )
    ).

%% string_trim(+String, -Trimmed)
%% Remove leading/trailing whitespace and CR
string_trim(String, Trimmed) :-
    split_string(String, "", " \t\r\n", [Trimmed|_]),
    !.
string_trim(_, "").

%% write_jsonrpc_response(+Stream, +Response)
%%
%% Writes a JSON-RPC response as a single line of JSON followed by a newline.
%% MCP stdio transport uses newline-delimited JSON.

write_jsonrpc_response(Stream, Response) :-
    with_output_to(string(JsonString),
        json_write_dict(current_output, Response, [width(0)])
    ),
    format(Stream, "~w\n", [JsonString]),
    flush_output(Stream).

%% write_jsonrpc_error(+Stream, +Id, +Code, +Message)
%%
%% Writes a JSON-RPC error response

write_jsonrpc_error(Stream, Id, Code, Message) :-
    Response = _{
        jsonrpc: "2.0",
        id: Id,
        error: _{
            code: Code,
            message: Message
        }
    },
    write_jsonrpc_response(Stream, Response).

%% write_jsonrpc_notification(+Stream, +Method, +Params)
%%
%% Writes a JSON-RPC notification (no id, no response expected)

write_jsonrpc_notification(Stream, Method, Params) :-
    Notification = _{
        jsonrpc: "2.0",
        method: Method,
        params: Params
    },
    write_jsonrpc_response(Stream, Notification).

%% jsonrpc_error_response(+Id, +Code, +Message, -Response)
%%
%% Creates a JSON-RPC error response dict

jsonrpc_error_response(Id, Code, Message, Response) :-
    Response = _{
        jsonrpc: "2.0",
        id: Id,
        error: _{
            code: Code,
            message: Message
        }
    }.

%% log_error(+Format, +Args)
%%
%% Logs an error to stderr

log_error(Format, Args) :-
    format(user_error, Format, Args),
    nl(user_error),
    flush_output(user_error).
